api/client/service.proto

```diff
@@ -34,10 +36,58 @@ message ExecutionResponse {
   google.rpc.Status error_detail = 7;
 }
 
+// Used to request a RequestStreamResponse.
+message RequestStreamRequest {
+  // Amount of RequestStreamResponses that should be yielded.
+  // Used to implement basic flow control by the client. Once AsyncClientImpl gets
+  // a way to apply backpressure this could probably be dropped.
+  uint64 amount = 1;
+}
+
+message RequestStreamResponse {
+  // Specifies what the requests look like. Will be merged with the client's configuration.
+  RequestSpecifier request_specifier = 1;
+  // Response-level expectations associated to the above request specification.
+  Expectations expectations = 2;
+}
+
+message RequestSpecifier {
+  // method/path/uri/authority/content_length can all be expressed in request headers.
+  // so technically they are redundant. In case of ambiguity headers may end up being duplicated.
+  // Note that as of today, the target upstream/cluster that will be used to replay traffic to
+  // fixes the destination address/port
+  // Request method
+  google.protobuf.StringValue method = 1;
+  // Request URI
+  google.protobuf.StringValue uri = 2;
+  // Request authority
+  google.protobuf.StringValue authority = 3;
+  // Request headers
+  envoy.api.v2.core.HeaderMap headers = 4;
+  // Request content length. The client will transfer the number of bytes specified here in
+  // for the request body.
+  google.protobuf.UInt32Value content_length = 5;
+  // TODO(oschaaf): nice to have
+  // google.protobuf.StringValue sni_hostname = 10;
+}
+
+// Response-level expectations. If set, any unmet expectations will result in stats counter
+// increases in the future.
+message Expectations {
+  // Expected response status code
+  google.protobuf.UInt32Value response_code = 1;
+  // Expected response body content length
+  google.protobuf.UInt32Value content_length = 2;
```

Comment:
Haha, for some reason I was thinking we had to commit to the ultimate interface up front, not realizing we would immediately have to implement everything that was declared

Reply:
> Haha, for some reason I was thinking we had to commit to the ultimate interface up front, not realizing we would immediately have to implement everything that was declared

That is an option. As far as I am concerned, we can iterate here in this PR or go and try to land the next steps in follow ups. Or we can go hybrid and try to shape this PR to the final state we want, and then split out sub PRs for in depth review when we are ready for that. But I stopped expanding this PR  earlier when I realized this change was getting to a complexity and size where it would be challenging to assess/land (even when already familiar with the code base).

Reply:
Filed #260 to propose adding expected headers and body later. We definitely want to keep this initial PR manageable to be able to release something sooner.